File: priv/combined_metrics/samples/variable_naming/loop_var_is_single_letter/good/matrix.js
// Matrix operations using single-letter loop counters.
// GOOD: numeric loop counters use i, j, k; element variables use descriptive names.

function multiply(matrixA, matrixB) {
  const rowCount = matrixA.length;
  const colCount = matrixB[0].length;
  const innerCount = matrixB.length;
  const result = [];

  for (let i = 0; i < rowCount; i++) {
    result[i] = [];
    for (let j = 0; j < colCount; j++) {
      let sum = 0;
      for (let k = 0; k < innerCount; k++) {
        sum += matrixA[i][k] * matrixB[k][j];
      }
      result[i][j] = sum;
    }
  }

  return result;
}

function transpose(matrix) {
  const rowCount = matrix.length;
  const colCount = matrix[0].length;
  const result = [];

  for (let j = 0; j < colCount; j++) {
    result[j] = [];
    for (let i = 0; i < rowCount; i++) {
      result[j][i] = matrix[i][j];
    }
  }

  return result;
}

function sumDiagonal(matrix) {
  let total = 0;
  for (let i = 0; i < matrix.length; i++) {
    total += matrix[i][i];
  }
  return total;
}

function fill(rows, cols, value) {
  const result = [];
  for (let i = 0; i < rows; i++) {
    result[i] = [];
    for (let j = 0; j < cols; j++) {
      result[i][j] = (i + 1) * (j + 1) * value;
    }
  }
  return result;
}

function findMaxInRow(matrix) {
  return matrix.map(row => {
    let max = row[0];
    for (let i = 1; i < row.length; i++) {
      if (row[i] > max) max = row[i];
    }
    return max;
  });
}

function rotate90(matrix) {
  const size = matrix.length;
  const result = [];

  for (let j = 0; j < size; j++) {
    result[j] = [];
    for (let i = size - 1; i >= 0; i--) {
      result[j][size - 1 - i] = matrix[i][j];
    }
  }

  return result;
}

function flatten(matrix) {
  const result = [];
  for (let i = 0; i < matrix.length; i++) {
    for (let j = 0; j < matrix[i].length; j++) {
      result.push(matrix[i][j]);
    }
  }
  return result;
}

module.exports = { multiply, transpose, sumDiagonal, fill, findMaxInRow, rotate90, flatten };

File: priv/combined_metrics/samples/variable_naming/loop_var_is_single_letter/good/matrix.rb
# Matrix operations using single-letter loop counters.
# GOOD: numeric loop counters use i, j, k; element variables use descriptive names.

class MatrixGood
  def multiply(matrix_a, matrix_b)
    row_count = matrix_a.length
    col_count = matrix_b[0].length
    inner_count = matrix_b.length
    result = Array.new(row_count) { Array.new(col_count, 0) }

    (0...row_count).each do |i|
      (0...col_count).each do |j|
        (0...inner_count).each do |k|
          result[i][j] += matrix_a[i][k] * matrix_b[k][j]
        end
      end
    end

    result
  end

  def transpose(matrix)
    row_count = matrix.length
    col_count = matrix[0].length
    result = Array.new(col_count) { Array.new(row_count) }

    (0...col_count).each do |j|
      (0...row_count).each do |i|
        result[j][i] = matrix[i][j]
      end
    end

    result
  end

  def sum_diagonal(matrix)
    total = 0
    (0...matrix.length).each do |i|
      total += matrix[i][i]
    end
    total
  end

  def fill(rows, cols, value)
    result = []
    (1..rows).each do |i|
      row = []
      (1..cols).each do |j|
        row << i * j * value
      end
      result << row
    end
    result
  end

  def find_max_in_row(matrix)
    matrix.map do |row|
      max = row[0]
      (1...row.length).each do |i|
        max = row[i] if row[i] > max
      end
      max
    end
  end

  def rotate_90(matrix)
    size = matrix.length
    result = Array.new(size) { Array.new(size) }

    (0...size).each do |j|
      (0...size).each do |i|
        result[j][size - 1 - i] = matrix[i][j]
      end
    end

    result
  end

  def flatten(matrix)
    result = []
    (0...matrix.length).each do |i|
      (0...matrix[i].length).each do |j|
        result << matrix[i][j]
      end
    end
    result
  end
end
